### ISSUE TITLE
Master branch is not setting appropriate version after PR'ing releases/ branch into it.

### ISSUE THREAD

**Issue by AndrewMcClelland** [GitTools/GitVersion#2490]

**Comment by asbjornu** [comment]
GitVerison can't base the version number of `master` on the name of another branch, since that branch is going to be deleted in the future. Git does not preserve branch names alongside the commits. The branches point to commits, not the other way around.

So even though it would be possible for GitVersion to generate the version number `1.5.0` on `master` on the commit `release/1.5.0` being merged into `master`, that's only possible while `release/1.5.0` still exists. Fast forward 2 years, GitVersion still needs to be able to mint version `1.5.0` from the same merge-commit, but since `release/1.5.0` is no longer existent, it's impossible for GitVersion to do so.

**Comment by AndrewMcClelland** [comment]
@asbjornu Thanks for the reply! Understood. 

Is the recommended flow then to merge `releases/1.5.0` into 'master', then manually tag `master` with `1.5.0`? Or is there a more automated way to have `master` version set to `1.5.0` upon PR completion?

For my understanding, why does this flow work when I do it locally? Is it because I merge the `releases/1.5.0` branch rather than PR it, so there's no temporary `pullrequest/` branch?

**Comment by timmyWelling** [comment]
The difference between the local test and ADO is likely the wording of the merge commit message. ADO is probably adding some prefix that disrupts GitVersion's default regex used to parse the branch name from the merge commit message. Like @asbjornu mentioned the branch isn't preserved, so GitVersion parses the merge commit to figure out where a branch merge occurred (e.g. Merged releases/1.5.0 into master). If the branch name is successfully parsed, the correct version should be calculated automatically. 

Check this documentation for an example: https://gitversion.net/docs/configuration#merge-message-formats

**Comment by AndrewMcClelland** [comment]
@timmyWelling Thanks for the information - I checked ADO and the commit messages associated with a squashed commit PR follow the format `Merged PR ######: PR Commit Message` whereas the local merge (where the `releases/1.5.0` merge to `master` sets the `master` version correctly to `1.5.0`) follows the format `Merge branch 'releases/1.5.0'`.

From what I've read of `merge-message-formats`, it seems like we can add something to the configuration file (global, not branch level (?)) that changes the commit message of a PR to a format we specify.

For example. I've added this to my `GitVersion.yml`:

```
merge-message-formats:
    ado: '^Merged (?:PR (?<PullRequestNumber>\d+)): Merge (?<SourceBranch>.+) to (?<TargetBranch>.+)'
```

I expect this would take ADO's PR commit message `Merged PR ######: PR Commit Message` and change it to something like `Merge releases/1.5.0 to master`. Is that correct?

I went ahead and PR'd the `merge-message-formats` change into develop, however the ADO PR squash commit remains the same format as `Merged PR ######: PR Commit Message`, so the version doesn't change. Do I have a formatting type somewhere in the `merge-message-formats` portion? Or am I misunderstanding this parameters functionality?

**Comment by AndrewMcClelland** [comment]
**Update**

Upon looking through [PR #1488](https://github.com/GitTools/GitVersion/pull/1488) and issue #1628 I have a better understanding of what the `merge-message-formats` does and how it works. However, I still haven't seen it work for me in ADO. The following is my setup and flow:

- `master` is currently at `1.0.1+3`
- branch off of `develop`, add the `merge-message-formats` parameter to the `GitVersion.yml`
  - ```
    merge-message-formats: {
      tfs: '^Merged (?:PR (?<PullRequestNumber>\d+)): Merge (?<SourceBranch>.+) to (?<TargetBranch>.+)'
    }
    ```
- Create a PR into `develop`, so now `develop` has the up to date `GitVersion.yml` with `merge-message-formats` included
- Create `releases/1.7.0` off of `develop` ... confirmed that new branch version is `1.7.0`
- Create a PR with a squash merge commit from `releases/1.7.0` into `master`. The PR merge commit message is `Merged PR 707677: Merge releases/1.7.0 to master` which matches the `merge-message-formats` regex and 3 group variables.
- Expect that `master` is now `1.7.0`. Instead, `master` is `1.0.1+4`

**Comment by timmyWelling** [comment]
You can test locally by manually recreating the same message used by TFS, and the output should tell you if it used your custom parser. My guess is remove the brackets in your config file.

**Comment by AndrewMcClelland** [comment]
@timmyWelling Sorry, not sure I understand what you mean by "the output should tell you if it used your custom parser"?

Will try removing brackets! Although they are included in the docs [here](https://gitversion.readthedocs.io/en/latest/input/docs/configuration/#global-configuration)

**Comment by AndrewMcClelland** [comment]
Also, does the `tfs` key here have any meaning? Or can I change it to `ado` with no issue?

**Comment by timmyWelling** [comment]
If you run GitVersion with the /output buildserver option, you will see the full version calculation logs. If the merge-message-parser regex matched on a commit, it will say 'Found commit x matching merge message format:tfs'. I believe you can change to ado.

**Comment by AndrewMcClelland** [comment]
@timmyWelling 

Ran the following steps locally, where `GitVersion.yml` has the same `merge-message-format`:
```
merge-message-formats:
    tfs: '^Merged (?:PR (?<PullRequestNumber>\d+)): Merge (?<SourceBranch>.+) to (?<TargetBranch>.+)'
```

- On `master` branch, version is `1.0.1+5`
- On `develop` branch, version is `1.0.1-alpha12`
- Checkout `releases/1.9.0` based on `develop`
- On `releases/1.9.0`, version is `1.9.0-beta.1`
- Checkout `master` and merge releases branch into master via `git merge releases/1.9.0 -m "Merged PR 999999: Merge releases/1.9.0 to master"` and got the output `Merge made by the 'recursive' strategy.`
- `dotnet-gitversion` on `master` now shows version `1.9.0` correctly
- `dotnet-gitversion /output buildserver` on `master` shows 
```
INFO [01/05/21 10:58:53:29] Working directory: REPO_NAME
INFO [01/05/21 10:58:53:33] Project root is: REPO_NAME\
INFO [01/05/21 10:58:53:33] DotGit directory is: REPO_NAME\.git
INFO [01/05/21 10:58:53:36] Begin: Loading version variables from disk cache
  INFO [01/05/21 10:58:53:36] Begin: Deserializing version variables from cache file REPO_NAME\.git\gitversion_cache\544F2465B57A85843EBA23BB6445F28E20C6C747.yml
  INFO [01/05/21 10:58:53:42] End: Deserializing version variables from cache file REPO_NAME\.git\gitversion_cache\544F2465B57A85843EBA23BB6445F28E20C6C747.yml (Took: 54.28ms)
  INFO [01/05/21 10:58:53:42] End: Loading version variables from disk cache (Took: 57.04ms)
  INFO [01/05/21 10:58:53:43] Done writing
```

As you can see, no mention of 'Found commit x matching merge message format:tfs' there, even though `master` was correctly updated to `1.9.0` after a merge with ADO PR squash commit message.

**Comment by AndrewMcClelland** [comment]
Additional info - this the output of `GitVersion` task in ADO for PR that was made into `master` from `releases/1.8.0`:

```
#1.0.1-PullRequest707750.17 Merge releases/1.8.0 to master
```
```
##[section]Starting: GitVersion
==============================================================================
Task         : GitVersion Task
Description  : Easy Semantic Versioning (http://semver.org) for projects using Git
Version      : 5.0.1
Author       : GitVersion Contributors
Help         : See the [documentation](http://gitversion.readthedocs.org/en/latest/) for help
==============================================================================
[command]"C:\Program Files\dotnet\dotnet.exe" D:\a\_tasks\GitVersion_e5983830-3f75-11e5-82ed-81492570a08e\5.0.1\core\GitVersion.dll D:\a\1\s /output buildserver /nofetch /config D:\a\1\s\GitVersion.yml /updateassemblyinfo true
INFO [01/05/21 0:58:59:37] Working directory: D:\a\1\s
INFO [01/05/21 0:58:59:46] IsDynamicGitRepository: False
INFO [01/05/21 0:58:59:74] Returning Project Root from DotGitDirectory: D:\a\1\s\.git - D:\a\1\s
INFO [01/05/21 0:58:59:74] Running on Windows.
INFO [01/05/21 0:58:59:75] Applicable build agent found: 'VsoAgent'.
INFO [01/05/21 0:58:59:75] Branch from build environment: refs/pull/707750/merge
INFO [01/05/21 0:58:59:75] Begin: Normalizing git directory for branch 'refs/pull/707750/merge'
  INFO [01/05/21 0:58:59:85] One remote found (origin -> 'https://XXXX.visualstudio.com/XXXX/_git/XXXX').
  INFO [01/05/21 0:58:59:86] Skipping fetching, if GitVersion does not calculate your version as expected you might need to allow fetching or use dynamic repositories
  INFO [01/05/21 0:58:59:87] Creating local branch refs/heads/pull/707750/merge pointing at 0f3a4775ddf91a33844feb645573fc0c73a64366
  INFO [01/05/21 0:58:59:91] Creating local branch from remote tracking 'refs/remotes/origin/0f3a4775ddf91a33844feb645573fc0c73a64366'.
  INFO [01/05/21 0:58:59:94] Creating local branch from remote tracking 'refs/remotes/origin/develop'.
  INFO [01/05/21 0:58:59:95] Creating local branch from remote tracking 'refs/remotes/origin/master'.
  INFO [01/05/21 0:58:59:96] Creating local branch from remote tracking 'refs/remotes/origin/releases/0.0.1'.
  INFO [01/05/21 0:58:59:96] Creating local branch from remote tracking 'refs/remotes/origin/releases/0.0.2'.
  INFO [01/05/21 0:58:59:97] Creating local branch from remote tracking 'refs/remotes/origin/releases/1.1.0'.
  INFO [01/05/21 0:58:59:97] Creating local branch from remote tracking 'refs/remotes/origin/releases/1.2.0'.
  INFO [01/05/21 0:58:59:98] Creating local branch from remote tracking 'refs/remotes/origin/releases/1.5.0'.
  INFO [01/05/21 0:58:59:98] Creating local branch from remote tracking 'refs/remotes/origin/releases/1.7.0'.
  INFO [01/05/21 0:58:59:99] Creating local branch from remote tracking 'refs/remotes/origin/releases/1.8.0'.
  INFO [01/05/21 0:59:00:00] HEAD points at branch 'refs/heads/pull/707750/merge'.
  INFO [01/05/21 0:59:00:00] End: Normalizing git directory for branch 'refs/pull/707750/merge' (Took: 253.81ms)
  INFO [01/05/21 0:59:00:00] IsDynamicGitRepository: False
  INFO [01/05/21 0:59:00:00] Returning Project Root from DotGitDirectory: D:\a\1\s\.git - D:\a\1\s
  INFO [01/05/21 0:59:00:00] Project root is: D:\a\1\s
  INFO [01/05/21 0:59:00:00] DotGit directory is: D:\a\1\s\.git
  INFO [01/05/21 0:59:00:19] IsDynamicGitRepository: False
  INFO [01/05/21 0:59:00:19] Returning Project Root from DotGitDirectory: D:\a\1\s\.git - D:\a\1\s
  INFO [01/05/21 0:59:00:20] Begin: Loading version variables from disk cache
    INFO [01/05/21 0:59:00:20] Cache file D:\a\1\s\.git\gitversion_cache\80E8AF79CD94331A9ABC643E11C180B6CFA158E6.yml not found.
  INFO [01/05/21 0:59:00:20] End: Loading version variables from disk cache (Took: 0.90ms)
  INFO [01/05/21 0:59:00:20] IsDynamicGitRepository: False
  INFO [01/05/21 0:59:00:20] Returning Project Root from DotGitDirectory: D:\a\1\s\.git - D:\a\1\s
  INFO [01/05/21 0:59:00:95] Using latest commit on specified branch
  INFO [01/05/21 0:59:00:97] Begin: Attempting to inherit branch configuration from parent branch
    INFO [01/05/21 0:59:00:98] HEAD is merge commit, this is likely a pull request using develop as base
    INFO [01/05/21 0:59:00:99] Begin: Finding branch source of 'develop'
      INFO [01/05/21 0:59:01:02] Begin: Finding merge base between 'develop' and 'master'.
        INFO [01/05/21 0:59:01:02] Found merge base of 0e422ec0217de001e9c1788b29f57cc151dc2379
        INFO [01/05/21 0:59:01:04] Merge base of develop' and 'master is 0e422ec0217de001e9c1788b29f57cc151dc2379
      INFO [01/05/21 0:59:01:04] End: Finding merge base between 'develop' and 'master'. (Took: 18.54ms)
    INFO [01/05/21 0:59:01:05] End: Finding branch source of 'develop' (Took: 59.45ms)
    INFO [01/05/21 0:59:01:05] Begin: Getting branches containing the commit '0e422ec0217de001e9c1788b29f57cc151dc2379'.
      INFO [01/05/21 0:59:01:05] Trying to find direct branches.
      INFO [01/05/21 0:59:01:05] No direct branches found, searching through tracked branches.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/develop'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/develop' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/hotfix/readme'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/hotfix/readme' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/master'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/master' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/0.0.1'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/0.0.1' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/0.0.2'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/0.0.2' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.1.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.1.0' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.2.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.2.0' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.5.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.5.0' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.7.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.7.0' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.8.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.8.0' has a matching commit.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/readme'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/readme' has no matching commits.
    INFO [01/05/21 0:59:01:06] End: Getting branches containing the commit '0e422ec0217de001e9c1788b29f57cc151dc2379'. (Took: 11.61ms)
    INFO [01/05/21 0:59:01:06] Begin: Getting branches containing the commit '0f3a4775ddf91a33844feb645573fc0c73a64366'.
      INFO [01/05/21 0:59:01:06] Trying to find direct branches.
      INFO [01/05/21 0:59:01:06] No direct branches found, searching through tracked branches.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/develop'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/develop' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/hotfix/readme'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/hotfix/readme' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/master'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/master' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/0.0.1'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/0.0.1' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/0.0.2'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/0.0.2' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.1.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.1.0' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.2.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.2.0' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.5.0'.
      INFO [01/05/21 0:59:01:06] The branch 'origin/releases/1.5.0' has no matching commits.
      INFO [01/05/21 0:59:01:06] Searching for commits reachable from 'origin/releases/1.7.0'.
      INFO [01/05/21 0:59:01:07] The branch 'origin/releases/1.7.0' has no matching commits.
      INFO [01/05/21 0:59:01:07] Searching for commits reachable from 'origin/releases/1.8.0'.
      INFO [01/05/21 0:59:01:07] The branch 'origin/releases/1.8.0' has no matching commits.
      INFO [01/05/21 0:59:01:07] Searching for commits reachable from 'origin/releases/readme'.
      INFO [01/05/21 0:59:01:07] The branch 'origin/releases/readme' has no matching commits.
    INFO [01/05/21 0:59:01:07] End: Getting branches containing the commit '0f3a4775ddf91a33844feb645573fc0c73a64366'. (Took: 4.27ms)
    INFO [01/05/21 0:59:01:07] Found possible parent branches: origin/develop, origin/master, origin/releases/0.0.1, origin/releases/0.0.2, origin/releases/1.1.0, origin/releases/1.2.0, origin/releases/1.5.0, origin/releases/1.7.0, origin/releases/1.8.0
    WARN [01/05/21 0:59:01:07] Failed to inherit Increment branch configuration, ended up with: origin/develop, origin/master, origin/releases/0.0.1, origin/releases/0.0.2, origin/releases/1.1.0, origin/releases/1.2.0, origin/releases/1.5.0, origin/releases/1.7.0, origin/releases/1.8.0

Falling back to develop branch config
  INFO [01/05/21 0:59:01:07] End: Attempting to inherit branch configuration from parent branch (Took: 98.11ms)
  INFO [01/05/21 0:59:01:08] Running against branch: pull/707750/merge (0f3a4775ddf91a33844feb645573fc0c73a64366)
  INFO [01/05/21 0:59:01:08] Begin: Calculating base versions
    INFO [01/05/21 0:59:01:09] Fallback base version: 0.1.0 with commit count source 0e422ec0217de001e9c1788b29f57cc151dc2379 (Incremented: None)
    INFO [01/05/21 0:59:01:14] NextVersion in GitVersion configuration file: 1.0.0 with commit count source External Source (Incremented: None)
    INFO [01/05/21 0:59:01:16] Git tag 'v1.0.0': 1.0.0 with commit count source 0e422ec0217de001e9c1788b29f57cc151dc2379 (Incremented: 1.0.1)
    INFO [01/05/21 0:59:01:24] Found multiple base versions which will produce the same SemVer (1.0.1), taking oldest source for commit counting (Git tag 'v1.0.0')
    INFO [01/05/21 0:59:01:25] Base version used: Git tag 'v1.0.0': 1.0.0 with commit count source 0e422ec0217de001e9c1788b29f57cc151dc2379 (Incremented: 1.0.1)
  INFO [01/05/21 0:59:01:25] End: Calculating base versions (Took: 162.84ms)
  INFO [01/05/21 0:59:01:26] 17 commits found between 0e422ec0217de001e9c1788b29f57cc151dc2379 and 0f3a4775ddf91a33844feb645573fc0c73a64366
  INFO [01/05/21 0:59:01:26] Begin: Getting version tags from branch 'refs/heads/pull/707750/merge'.
  INFO [01/05/21 0:59:01:26] End: Getting version tags from branch 'refs/heads/pull/707750/merge'. (Took: 3.67ms)
  INFO [01/05/21 0:59:01:31] Begin: Creating dictionary
  INFO [01/05/21 0:59:01:32] End: Creating dictionary (Took: 7.72ms)
  INFO [01/05/21 0:59:01:33] Begin: Storing version variables to cache file D:\a\1\s\.git\gitversion_cache\80E8AF79CD94331A9ABC643E11C180B6CFA158E6.yml
  INFO [01/05/21 0:59:01:46] End: Storing version variables to cache file D:\a\1\s\.git\gitversion_cache\80E8AF79CD94331A9ABC643E11C180B6CFA158E6.yml (Took: 125.68ms)
  INFO [01/05/21 0:59:01:51] Applicable build agent found: 'VsoAgent'.
Executing GenerateSetVersionMessage for 'VsoAgent'.
Executing GenerateBuildLogOutput for 'VsoAgent'.
  INFO [01/05/21 0:59:01:53] Updating assembly info files
  INFO [01/05/21 0:59:01:53] Found 0 files
##[section]Async Command Start: Update Build Number
Update build number to 1.0.1-PullRequest707750.17 for build 5005243
##[section]Async Command End: Update Build Number
##[section]Finishing: GitVersion
```
